pkg/tunneler/listener.go

```diff
@@ -0,0 +1,290 @@
+		return nil, err
+	}
+	if res.StatusCode != 200 {
+		klog.V(5).Infof("Status code %d on request %v, retry %d", res.StatusCode, ln.url)
+		return nil, fmt.Errorf("status code %d", res.StatusCode)
+	}
+
+	conn := rwconn.NewConn(res.Body, pw)
+	return conn, nil
+}
+
+func (ln *Listener) grabConn() {
+	// create a new connection
+	c, err := ln.dial()
+	if err != nil {
+		klog.V(5).Infof("Can not create connection %v", err)
+		ln.sendMessage(controlMsg{Command: "pickup-failed", ConnPath: "", Err: err.Error()})
+		return
+	}
+
+	// send the connection to the listener
+	select {
+	case <-ln.donec:
+		return
+	default:
+		select {
+		case ln.connc <- c:
+		case <-ln.donec:
+			return
+		}
+	}
+
+}
+
+// Accept blocks and returns a new connection, or an error.
+func (ln *Listener) Accept() (net.Conn, error) {
+	c, ok := <-ln.connc
+	if !ok {
+		ln.mu.Lock()
+		err, closed := ln.readErr, ln.closed
+		ln.mu.Unlock()
+		if err != nil && !closed {
+			return nil, fmt.Errorf("tunneler: Listener closed; %w", err)
+		}
+		return nil, ErrListenerClosed
+	}
+	klog.V(5).Infof("Accept connection")
+	return c, nil
+}
+
+// ErrListenerClosed is returned by Accept after Close has been called.
+var ErrListenerClosed = errors.New("tunneler: Listener closed")
+
+// Close closes the Listener, making future Accept calls return an
+// error.
+func (ln *Listener) Close() error {
+	ln.mu.Lock()
+	defer ln.mu.Unlock()
+	if ln.closed {
+		return nil
+	}
+	ln.closed = true
+	close(ln.connc)
+	close(ln.donec)
+	ln.sc.Close()
+	return nil
+}
+
+// Addr returns a dummy address. This exists only to conform to the
+// net.Listener interface.
+func (ln *Listener) Addr() net.Addr { return connAddr{} }
+
+// configureHTTP2Transport enable ping to avoid issues with stale connections
+func configureHTTP2Transport(client *http.Client) error {
+	t, ok := client.Transport.(*http.Transport)
+	if !ok {
+		// can't get the transport it will fail later if not http2 supported
+		return nil
+	}
+
+	if t.TLSClientConfig == nil {
+		return fmt.Errorf("only TLS supported")
+	}
+
+	for _, v := range t.TLSClientConfig.NextProtos {
+		// http2 already configured
+		if v == "h2" {
+			return nil
+		}
+	}
+
+	t2, err := http2.ConfigureTransports(t)
+	if err != nil {
+		return err
+	}
+
+	t2.ReadIdleTimeout = time.Duration(30) * time.Second
+	t2.PingTimeout = time.Duration(15) * time.Second
+	return nil
+}
+
+// syncerTunnelURL builds the destination url with the Dialer expected format of the URL
+func syncerTunnelURL(host string, id string) (string, error) {
+	if id == "" {
+		return "", fmt.Errorf("id can not be empty")
+	}
+	hostURL, err := url.Parse(host)
+	if err != nil || hostURL.Scheme != "https" || hostURL.Host == "" {
+		return "", fmt.Errorf("wrong url format, expected https://host<:port>/<path>: %w", err)
+	}
+	host = strings.Trim(host, "/")
+	return host + defaultTunnelPathPrefix + "/" + cmdTunnelConnect + "/" + id, nil
```

Comment:
the URL should be `https://host/services/tunnels/clusters/ws/...`, not services behind clusters.

Reply:
while at it, would even rename it to `the URL should be `https://host/services/syncer-tunnels/clusters/ws/...`

Reply:
And to make it complete: `https://host/services/syncer-tunnels/clusters/<ws>/apis/workload.kcp.dev/v1alpha1/synctargets/<name>/tunnel`

Reply:
if I do that , don't I miss the cluster name from the context added in `WithClusterScope`?

```
func WithClusterScope(apiHandler http.Handler) http.HandlerFunc {
	return func(w http.ResponseWriter, req *http.Request) {
		var clusterName logicalcluster.Name
		if path := req.URL.Path; strings.HasPrefix(path, "/clusters/") {
			path = strings.TrimPrefix(path, "/clusters/")

			i := strings.Index(path, "/")
```

Reply:
yes, under `/services` you have to extract the cluster manually yourself. That's expected.

Reply:
Note: with a follow-up this will only be a problem for the syncer tunnel side, not the kubectl/client side. The later will move to  `/clusters/<ws>/api/v1/namespace/<ns>/pods/<name>/log`.

Reply:
this is still open

Reply:
fixed, now